jax/_src/nn/functions.py

```diff
@@ -765,3 +767,178 @@ def hard_silu(x: ArrayLike) -> Array:
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    is_causal: bool = False,
+    implementation: str | None = None) -> Array | tuple[Array, Array]:
+  r"""Scaled dot product attention function.
+
+  Computes the attention function on Query, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+          :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. It is a boolean mask
+          where `True` indicates the element should take part in attention. For
+          an additive mask, users should pass it to `bias`. The shape is
+          broadcastable to :code:`(BNTS)`.
+    scale: scale for the logits. If None, the scale will be set to 1 divided by
+           the square root of query's head dimension (i.e. H).
+    is_causal: If true, causal attention will be applied. Note, some
+               implementations like `xla` will generate a mask tensor and apply
+               it to the logits, but other implementations like `cudnn` will
+               avoid computing the unmasked regions.
+    implementation: A string to control which implementation backend to use.
+                    Supported strings are `xla`, `cudnn` (cuDNN flash
+                    attention). It defaults to `None`, which will automatically
+                    select the best available backend.
+
+  Returns:
+    An array of the attention output with the same shape of :code:`query`.
+
+  Raises:
+    ValueError: if the `implementation` is not a supported option from
+                ['xla', 'cudnn', None].
+
+  """
+  def _assert_has_shape(t: ArrayLike, shape: Sequence[int]) -> None:
+    assert t.ndim == len(shape)
+    value_str1 = f't.shape={t.shape}'
+    value_str2 = f'shape={shape}'
+    for i in range(t.ndim):
+      if shape[i] != -1:
+        assert (
+            t.shape[i] == shape[i]
+        ), f'Shapes are not equal {value_str1} vs {value_str2}'
+    
+  B, S, N, H = key.shape
+  _assert_has_shape(value, [B, S, N, H])
+  _assert_has_shape(query, [B, -1, N, H])
+  T = query.shape[1]
+  scale_val = (1.0 / np.sqrt(H)) if scale is None else scale
+  assert query.dtype == key.dtype == value.dtype
+  if mask is not None:
+    assert mask.dtype == query.dtype and mask.dtype == jnp.bool_
```

Comment:
Masks must be bools. Also: raise a value error than use an assert for checking user inputs.

Reply:
Done.